src/vs/platform/files/node/diskFileSystemProvider.ts

```diff
@@ -172,13 +175,57 @@ export class DiskFileSystemProvider extends AbstractDiskFileSystemProvider imple
 
 	//#region File Reading/Writing
 
-	async readFile(resource: URI): Promise<Uint8Array> {
+	private readonly writeBarriers = new ResourceMap<Barrier>(resource => extUriBiasedIgnorePathCase.getComparisonKey(resource));
+
+	private async createWriteBarrier(resource: URI): Promise<IDisposable> {
+
+		// Await pending barriers for resource
+		// It is possible for a new barrier being
+		// added right after opening, so we have
+		// to loop over barriers until no barrier
+		// remains.
+		let existingWriteBarrier: Barrier | undefined = undefined;
```

Comment:
I understand the comment about possibility of having mutliple locks per file, but I think this cannot happen right? For eg., each write (or read) always locks a resource and other write/read on same resource cannot lock it again? Instead these operations have to wait until the lock is released right?